// nativescript/lib/common/services/project-files-provider-base.js
"use strict";
const path = require("path");
const util = require("util");
const constants_1 = require("../constants");
class ProjectFilesProviderBase {
    constructor($mobileHelper, $options) {
        this.$mobileHelper = $mobileHelper;
        this.$options = $options;
    }
    getPreparedFilePath(filePath) {
        let projectFileInfo = this.getProjectFileInfo(filePath, "");
        return path.join(path.dirname(filePath), projectFileInfo.onDeviceFileName);
    }
    getProjectFileInfo(filePath, platform, projectFilesConfig) {
        let parsed = this.parseFile(filePath, this.$mobileHelper.platformNames, platform || "");
        let basicConfigurations = [constants_1.Configurations.Debug.toLowerCase(), constants_1.Configurations.Release.toLowerCase()];
        if (!parsed) {
            let validValues = basicConfigurations.concat(projectFilesConfig && projectFilesConfig.additionalConfigurations || []), value = projectFilesConfig && projectFilesConfig.configuration || basicConfigurations[0];
            parsed = this.parseFile(filePath, validValues, value);
        }
        return parsed || {
            filePath: filePath,
            onDeviceFileName: path.basename(filePath),
            shouldIncludeFile: true
        };
    }
    parseFile(filePath, validValues, value) {
        let regex = util.format("^(.+?)[.](%s)([.].+?)$", validValues.join("|"));
        let parsed = filePath.match(new RegExp(regex, "i"));
        if (parsed) {
            return {
                filePath: filePath,
                onDeviceFileName: path.basename(parsed[1] + parsed[3]),
                shouldIncludeFile: parsed[2].toLowerCase() === value.toLowerCase()
            };
        }
        return null;
    }
}
exports.ProjectFilesProviderBase = ProjectFilesProviderBase;
