// nativescript/lib/commands/prepare.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
class PrepareCommand {
    constructor($options, $platformService, $projectData, $platformCommandParameter) {
        this.$options = $options;
        this.$platformService = $platformService;
        this.$projectData = $projectData;
        this.$platformCommandParameter = $platformCommandParameter;
        this.allowedParameters = [this.$platformCommandParameter];
        this.$projectData.initializeProjectData();
    }
    execute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const appFilesUpdaterOptions = { bundle: this.$options.bundle, release: this.$options.release };
            yield this.$platformService.preparePlatform(args[0], appFilesUpdaterOptions, this.$options.platformTemplate, this.$projectData, { provision: this.$options.provision, sdk: this.$options.sdk });
        });
    }
    canExecute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.$platformCommandParameter.validate(args[0])) && (yield this.$platformService.validateOptions(this.$options.provision, this.$projectData, args[0]));
        });
    }
}
exports.PrepareCommand = PrepareCommand;
$injector.registerCommand("prepare", PrepareCommand);
