// nativescript/lib/common/prompter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const prompt = require("inquirer");
const helpers = require("./helpers");
const readline = require("readline");
let MuteStream = require("mute-stream");
class Prompter {
    constructor() {
        this.muteStreamInstance = null;
    }
    dispose() {
        if (this.ctrlcReader) {
            this.ctrlcReader.close();
        }
    }
    get(schemas) {
        return __awaiter(this, void 0, void 0, function* () {
            let promptResult;
            try {
                promptResult = yield new Promise((resolve, reject) => {
                    this.muteStdout();
                    if (!helpers.isInteractive()) {
                        if (_.some(schemas, s => !s.default)) {
                            reject(new Error("Console is not interactive and no default action specified."));
                        }
                        else {
                            let result = {};
                            _.each(schemas, s => {
                                result[s.name] = s.default();
                            });
                            resolve(result);
                        }
                    }
                    else {
                        prompt.prompt(schemas, (result) => {
                            if (result) {
                                resolve(result);
                            }
                            else {
                                reject(new Error(`Unable to get result from prompt: ${result}`));
                            }
                        });
                    }
                });
            }
            finally {
                this.unmuteStdout();
            }
            return promptResult;
        });
    }
    getPassword(prompt, options) {
        return __awaiter(this, void 0, void 0, function* () {
            let schema = {
                message: prompt,
                type: "password",
                name: "password",
                validate: (value) => {
                    let allowEmpty = options && options.allowEmpty;
                    return (!allowEmpty && !value) ? "Password must be non-empty" : true;
                }
            };
            let result = yield this.get([schema]);
            return result.password;
        });
    }
    getString(prompt, options) {
        return __awaiter(this, void 0, void 0, function* () {
            let schema = {
                message: prompt,
                type: "input",
                name: "inputString",
                validate: (value) => {
                    let doesNotAllowEmpty = options && _.has(options, "allowEmpty") && !options.allowEmpty;
                    return (doesNotAllowEmpty && !value) ? `${prompt} must be non-empty` : true;
                },
                default: options && options.defaultAction
            };
            let result = yield this.get([schema]);
            return result.inputString;
        });
    }
    promptForChoice(promptMessage, choices) {
        return __awaiter(this, void 0, void 0, function* () {
            let schema = {
                message: promptMessage,
                type: "list",
                name: "userAnswer",
                choices: choices
            };
            let result = yield this.get([schema]);
            return result.userAnswer;
        });
    }
    confirm(prompt, defaultAction) {
        return __awaiter(this, void 0, void 0, function* () {
            let schema = {
                type: "confirm",
                name: "prompt",
                default: defaultAction,
                message: prompt
            };
            let result = yield this.get([schema]);
            return result.prompt;
        });
    }
    muteStdout() {
        if (helpers.isInteractive()) {
            process.stdin.setRawMode(true);
            this.muteStreamInstance = new MuteStream();
            this.muteStreamInstance.pipe(process.stdout);
            this.muteStreamInstance.mute();
            this.ctrlcReader = readline.createInterface({
                input: process.stdin,
                output: this.muteStreamInstance
            });
            this.ctrlcReader.on("SIGINT", process.exit);
        }
    }
    unmuteStdout() {
        if (helpers.isInteractive()) {
            process.stdin.setRawMode(false);
            if (this.muteStreamInstance) {
                this.cleanEventListeners(process.stdout);
                this.muteStreamInstance.unmute();
                this.muteStreamInstance = null;
                this.dispose();
            }
        }
    }
    cleanEventListeners(stream) {
        let memoryLeakEvents = [{
                eventName: "close",
                listenerName: "cleanup"
            }, {
                eventName: "error",
                listenerName: "onerror"
            }, {
                eventName: "drain",
                listenerName: "ondrain"
            }];
        _.each(memoryLeakEvents, (memoryleakEvent) => this.cleanListener(stream, memoryleakEvent.eventName, memoryleakEvent.listenerName));
    }
    cleanListener(stream, eventName, listenerName) {
        let eventListeners = process.stdout.listeners(eventName);
        let listenerFunction = _.find(eventListeners, (func) => func.name === listenerName);
        if (listenerFunction) {
            stream.removeListener(eventName, listenerFunction);
        }
    }
}
exports.Prompter = Prompter;
$injector.register("prompter", Prompter);
