// nativescript/lib/common/appbuilder/services/livesync/ios-livesync-service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const path = require("path");
const shell = require("shelljs");
let osenv = require("osenv");
const constants = require("../../../constants");
class IOSLiveSyncService {
    constructor(_device, $fs, $injector, $logger, $errors, $options, $iosDeviceOperations) {
        this._device = _device;
        this.$fs = $fs;
        this.$injector = $injector;
        this.$logger = $logger;
        this.$errors = $errors;
        this.$options = $options;
        this.$iosDeviceOperations = $iosDeviceOperations;
        this.$iosDeviceOperations.setShouldDispose(!this.$options.watch);
    }
    get $project() {
        return this.$injector.resolve("project");
    }
    get device() {
        return this._device;
    }
    refreshApplication(deviceAppData, localToDevicePaths) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.device.isEmulator) {
                let simulatorLogFilePath = path.join(osenv.home(), `/Library/Developer/CoreSimulator/Devices/${this.device.deviceInfo.identifier}/data/Library/Logs/system.log`);
                let simulatorLogFileContent = this.$fs.readText(simulatorLogFilePath) || "";
                let simulatorCachePath = path.join(osenv.home(), `/Library/Developer/CoreSimulator/Devices/${this.device.deviceInfo.identifier}/data/Containers/Data/Application/`);
                let regex = new RegExp(`^(?:.*?)${deviceAppData.appIdentifier}(?:.*?)${simulatorCachePath}(.*?)$`, "gm");
                let guid = "";
                while (true) {
                    let parsed = regex.exec(simulatorLogFileContent);
                    if (!parsed) {
                        break;
                    }
                    guid = parsed[1];
                }
                if (!guid) {
                    this.$errors.failWithoutHelp(`Unable to find application GUID for application ${deviceAppData.appIdentifier}. Make sure application is installed on Simulator.`);
                }
                let sourcePath = yield deviceAppData.getDeviceProjectRootPath();
                let destinationPath = path.join(simulatorCachePath, guid, constants.LiveSyncConstants.IOS_PROJECT_PATH);
                this.$logger.trace(`Transferring from ${sourcePath} to ${destinationPath}`);
                shell.cp("-Rf", path.join(sourcePath, "*"), destinationPath);
                yield this.device.applicationManager.restartApplication(deviceAppData.appIdentifier);
            }
            else {
                yield this.device.fileSystem.deleteFile("/Documents/AppBuilder/ServerInfo.plist", deviceAppData.appIdentifier);
                let notification = this.$project.projectData.Framework === constants.TARGET_FRAMEWORK_IDENTIFIERS.NativeScript ? "com.telerik.app.refreshApp" : "com.telerik.app.refreshWebView";
                const notificationData = {
                    deviceId: this.device.deviceInfo.identifier,
                    notificationName: notification,
                    commandType: constants.IOS_POST_NOTIFICATION_COMMAND_TYPE
                };
                yield this.$iosDeviceOperations.postNotification([notificationData]);
            }
        });
    }
    removeFiles(appIdentifier, localToDevicePaths) {
        return __awaiter(this, void 0, void 0, function* () {
            const devicePaths = localToDevicePaths.map(localToDevicePath => localToDevicePath.getDevicePath());
            for (let deviceFilePath of devicePaths) {
                yield this.device.fileSystem.deleteFile(deviceFilePath, appIdentifier);
            }
        });
    }
}
exports.IOSLiveSyncService = IOSLiveSyncService;
$injector.register("iosLiveSyncServiceLocator", { factory: IOSLiveSyncService });
