// nativescript/lib/services/platform-project-service-base.js
"use strict";
const events_1 = require("events");
class PlatformProjectServiceBase extends events_1.EventEmitter {
    constructor($fs, $projectDataService) {
        super();
        this.$fs = $fs;
        this.$projectDataService = $projectDataService;
    }
    getPluginPlatformsFolderPath(pluginData, platform) {
        return pluginData.pluginPlatformsFolderPath(platform);
    }
    getAllNativeLibrariesForPlugin(pluginData, platform, filter) {
        let pluginPlatformsFolderPath = this.getPluginPlatformsFolderPath(pluginData, platform), nativeLibraries = [];
        if (pluginPlatformsFolderPath && this.$fs.exists(pluginPlatformsFolderPath)) {
            let platformsContents = this.$fs.readDirectory(pluginPlatformsFolderPath);
            nativeLibraries = _(platformsContents)
                .filter(platformItemName => filter(platformItemName, pluginPlatformsFolderPath))
                .value();
        }
        return nativeLibraries;
    }
    getFrameworkVersion(runtimePackageName, projectDir) {
        return this.$projectDataService.getNSValue(projectDir, runtimePackageName).version;
    }
}
exports.PlatformProjectServiceBase = PlatformProjectServiceBase;
