// nativescript/lib/common/appbuilder/device-emitter.js
"use strict";
const events_1 = require("events");
class DeviceEmitter extends events_1.EventEmitter {
    constructor($androidDeviceDiscovery, $iOSDeviceDiscovery, $iOSSimulatorDiscovery, $deviceLogProvider, $companionAppsService) {
        super();
        this.$androidDeviceDiscovery = $androidDeviceDiscovery;
        this.$iOSDeviceDiscovery = $iOSDeviceDiscovery;
        this.$iOSSimulatorDiscovery = $iOSSimulatorDiscovery;
        this.$deviceLogProvider = $deviceLogProvider;
        this.$companionAppsService = $companionAppsService;
        this.initialize();
    }
    get companionAppIdentifiers() {
        if (!this._companionAppIdentifiers) {
            this._companionAppIdentifiers = this.$companionAppsService.getAllCompanionAppIdentifiers();
        }
        return this._companionAppIdentifiers;
    }
    initialize() {
        this.$androidDeviceDiscovery.on("deviceFound", (device) => {
            this.emit("deviceFound", device.deviceInfo);
            this.attachApplicationChangedHandlers(device);
            device.openDeviceLogStream();
        });
        this.$androidDeviceDiscovery.on("deviceLost", (device) => {
            this.emit("deviceLost", device.deviceInfo);
        });
        this.$iOSDeviceDiscovery.on("deviceFound", (device) => {
            this.emit("deviceFound", device.deviceInfo);
            this.attachApplicationChangedHandlers(device);
            device.openDeviceLogStream();
        });
        this.$iOSDeviceDiscovery.on("deviceLost", (device) => {
            this.emit("deviceLost", device.deviceInfo);
        });
        this.$iOSSimulatorDiscovery.on("deviceFound", (device) => {
            this.emit("deviceFound", device.deviceInfo);
            device.openDeviceLogStream();
            this.attachApplicationChangedHandlers(device);
        });
        this.$iOSSimulatorDiscovery.on("deviceLost", (device) => {
            this.emit("deviceLost", device.deviceInfo);
        });
        this.$deviceLogProvider.on("data", (identifier, data) => {
            this.emit('deviceLogData', identifier, data.toString());
        });
    }
    attachApplicationChangedHandlers(device) {
        device.applicationManager.on("applicationInstalled", (appIdentifier) => {
            this.emit("applicationInstalled", device.deviceInfo.identifier, appIdentifier);
            this.checkCompanionAppChanged(device, appIdentifier, "companionAppInstalled");
        });
        device.applicationManager.on("applicationUninstalled", (appIdentifier) => {
            this.emit("applicationUninstalled", device.deviceInfo.identifier, appIdentifier);
            this.checkCompanionAppChanged(device, appIdentifier, "companionAppUninstalled");
        });
        device.applicationManager.on("debuggableAppFound", (debuggableAppInfo) => {
            this.emit("debuggableAppFound", debuggableAppInfo);
        });
        device.applicationManager.on("debuggableAppLost", (debuggableAppInfo) => {
            this.emit("debuggableAppLost", debuggableAppInfo);
        });
        device.applicationManager.on("debuggableViewFound", (appIdentifier, debuggableWebViewInfo) => {
            this.emit("debuggableViewFound", device.deviceInfo.identifier, appIdentifier, debuggableWebViewInfo);
        });
        device.applicationManager.on("debuggableViewLost", (appIdentifier, debuggableWebViewInfo) => {
            this.emit("debuggableViewLost", device.deviceInfo.identifier, appIdentifier, debuggableWebViewInfo);
        });
        device.applicationManager.on("debuggableViewChanged", (appIdentifier, debuggableWebViewInfo) => {
            this.emit("debuggableViewChanged", device.deviceInfo.identifier, appIdentifier, debuggableWebViewInfo);
        });
    }
    checkCompanionAppChanged(device, applicationName, eventName) {
        let devicePlatform = device.deviceInfo.platform.toLowerCase();
        _.each(this.companionAppIdentifiers, (platformsCompanionAppIdentifiers, framework) => {
            if (applicationName === platformsCompanionAppIdentifiers[devicePlatform]) {
                this.emit(eventName, device.deviceInfo.identifier, framework);
                return false;
            }
        });
    }
}
exports.DeviceEmitter = DeviceEmitter;
$injector.register("deviceEmitter", DeviceEmitter);
