// nativescript/lib/common/services/plugins/npm-registry-plugins-source.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const plugins_source_base_1 = require("./plugins-source-base");
class NpmRegistryPluginsSource extends plugins_source_base_1.PluginsSourceBase {
    constructor($progressIndicator, $logger, $npmService) {
        super($progressIndicator, $logger);
        this.$progressIndicator = $progressIndicator;
        this.$logger = $logger;
        this.$npmService = $npmService;
    }
    get progressIndicatorMessage() {
        return "Searching for plugin in http://registry.npmjs.org.";
    }
    getPlugins(page, count) {
        return __awaiter(this, void 0, void 0, function* () {
            return page === 1 ? this.plugins : null;
        });
    }
    initializeCore(projectDir, keywords) {
        return __awaiter(this, void 0, void 0, function* () {
            let plugin = yield this.getPluginFromNpmRegistry(keywords[0]);
            this.plugins = plugin ? [plugin] : null;
        });
    }
    prepareScopedPluginName(plugin) {
        return plugin.replace("/", "%2F");
    }
    getPluginFromNpmRegistry(plugin) {
        return __awaiter(this, void 0, void 0, function* () {
            let dependencyInfo = this.$npmService.getDependencyInformation(plugin);
            let pluginName = this.$npmService.isScopedDependency(plugin) ? this.prepareScopedPluginName(dependencyInfo.name) : dependencyInfo.name;
            let result = yield this.$npmService.getPackageJsonFromNpmRegistry(pluginName, dependencyInfo.version);
            if (!result) {
                return null;
            }
            result.author = (result.author && result.author.name) || result.author;
            return result;
        });
    }
}
exports.NpmRegistryPluginsSource = NpmRegistryPluginsSource;
