// nativescript/lib/common/commands/help.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments)).next());
    });
};
const os_1 = require("os");
class HelpCommand {
    constructor($logger, $injector, $htmlHelpService, $staticConfig, $options) {
        this.$logger = $logger;
        this.$injector = $injector;
        this.$htmlHelpService = $htmlHelpService;
        this.$staticConfig = $staticConfig;
        this.$options = $options;
        this.enableHooks = false;
        this.allowedParameters = [];
    }
    canExecute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return true;
        });
    }
    execute(args) {
        return __awaiter(this, void 0, void 0, function* () {
            let topic = (args[0] || "").toLowerCase();
            let hierarchicalCommand = this.$injector.buildHierarchicalCommand(args[0], _.tail(args));
            if (hierarchicalCommand) {
                topic = hierarchicalCommand.commandName;
            }
            if (this.$options.help) {
                let help = yield this.$htmlHelpService.getCommandLineHelpForCommand(topic);
                if (this.$staticConfig.FULL_CLIENT_NAME) {
                    this.$logger.info(this.$staticConfig.FULL_CLIENT_NAME.green.bold + os_1.EOL);
                }
                this.$logger.printMarkdown(help);
            }
            else {
                yield this.$htmlHelpService.openHelpForCommandInBrowser(topic);
            }
        });
    }
}
exports.HelpCommand = HelpCommand;
$injector.registerCommand(["help", "/?"], HelpCommand);
